pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+		ensure!(asset != basic_asset, Error::<T>::WrongAssetToBuyout);
+
+		let (basic_asset_price, exchange_asset_price) = Self::fetch_prices((&basic_asset, &asset))?;
+
+		// Add fee to the basic asset price
+		let basic_asset_price_with_fee =
+			basic_asset_price * (FixedU128::from(T::SellFee::get()) + FixedU128::one());
+
+		let exchange_amount = Self::multiply_by_rational(
+			buyout_amount.saturated_into::<u128>(),
+			basic_asset_price_with_fee.into_inner(),
+			exchange_asset_price.into_inner(),
+		)
+		.map(|n| n.try_into().ok())
+		.flatten()
+		.ok_or(ArithmeticError::Overflow.into());
+
+		exchange_amount
+	}
+
+	/// Used for calculating buyout amount of basic asset user will get for exchange_amount of exchange asset
+	fn calc_buyout_amount(
+		asset: CurrencyIdOf<T>,
+		exchange_amount: BalanceOf<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let basic_asset = <T as orml_currencies::Config>::GetNativeCurrencyId::get();
+
+		ensure!(asset != basic_asset, Error::<T>::WrongAssetToBuyout);
+
+		let (basic_asset_price, exchange_asset_price) = Self::fetch_prices((&basic_asset, &asset))?;
+
+		// Add fee to the basic asset price
+		let basic_asset_price_with_fee =
+			basic_asset_price * (FixedU128::from(T::SellFee::get()) + FixedU128::one());
+
+		let buyout_amount = Self::multiply_by_rational(
+			exchange_amount.saturated_into::<u128>(),
+			exchange_asset_price.into_inner(),
+			basic_asset_price_with_fee.into_inner(),
+		)
+		.map(|b| b.try_into().ok())
+		.flatten()
+		.ok_or(ArithmeticError::Overflow.into());
+
+		buyout_amount
+	}
+
+	/// Used for splitting calculations of amount based on the input given
+	/// If user's call contains buyout amount, then exchange amount is calculated and viceversa
+	fn split_to_buyout_and_exchange(
+		asset: CurrencyIdOf<T>,
+		amount: Amount<BalanceOf<T>>,
+	) -> Result<(BalanceOf<T>, BalanceOf<T>), DispatchError> {
+		match amount {
+			Amount::Buyout(buyout_amount) => {
+				let exchange_amount = Self::calc_amount_to_exchange(asset, buyout_amount)?;
+				Ok((buyout_amount, exchange_amount))
+			},
+			Amount::Exchange(exchange_amount) => {
+				let buyout_amount = Self::calc_buyout_amount(asset, exchange_amount)?;
+				Ok((buyout_amount, exchange_amount))
+			},
+		}
+	}
+
+	fn do_buyout(
+		who: AccountIdOf<T>,
+		asset: CurrencyIdOf<T>,
+		amount: Amount<BalanceOf<T>>,
+	) -> DispatchResult {
+		Self::ensure_allowed_asset_for_buyout(&asset)?;
+
+		let basic_asset = <T as orml_currencies::Config>::GetNativeCurrencyId::get();
+		let (buyout_amount, exchange_amount) = Self::split_to_buyout_and_exchange(asset, amount)?;
+
+		Self::ensure_buyout_limit_not_exceeded(&who, buyout_amount)?;
+		let treasury_account_id = T::TreasuryAccount::get();
+
+		// Start exchanging
+		// Check for exchanging zero values and same accounts
+		if exchange_amount.is_zero() && buyout_amount.is_zero() || who == treasury_account_id {
+			return Ok(())
```

Comment:
Should we return an error here?

Reply:
Or maybe two different errors even

Reply:
Yes, we should definitely return 2 different errors 👍🏼